src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -191,15 +315,97 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            if (showFieldName) {
+                fieldDataList.add(fieldName);
+            }
+            if (showFieldType) {
+                String fieldType = getFieldType(fieldName, propertiesAsMap, index);
+                if (fieldType != null && !fieldType.isEmpty()) {
+                    fieldDataList.add(fieldType);
+                }
+            }
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    String getFieldType(String fieldName, Map<String, Object> propertiesAsMap, Index index) {
+        if (propertiesAsMap == null || index == null) {
+            return null;
+        }
+        // Attempt to get field type from cache
+        String fieldType = getFieldTypeFromCache(fieldName, index);
+
+        if (fieldType != null) {
+            return fieldType;
+        }
+
+        // Retrieve field type from mapping and cache it if found
+        fieldType = getFieldTypeFromProperties(fieldName, propertiesAsMap);
+
+        // Cache field type or NO_FIELD_TYPE_VALUE if not found
+        fieldTypeMap.computeIfAbsent(index, k -> new ConcurrentHashMap<>())
+            .putIfAbsent(fieldName, fieldType != null ? fieldType : NO_FIELD_TYPE_VALUE);
+
+        return fieldType;
+    }
+
+    String getFieldTypeFromProperties(String fieldName, Map<String, Object> propertiesAsMap) {
+        if (propertiesAsMap == null) {
+            return null;
+        }
+
+        String[] fieldParts = fieldName.split("\\.");
+        Map<String, ?> currentProperties = propertiesAsMap;
+
+        for (int depth = 0; depth < fieldParts.length; depth++) {
+            Object currentMapping = currentProperties.get(fieldParts[depth]);
+
+            if (currentMapping instanceof Map) {
+                Map<String, ?> currentMap = (Map<String, ?>) currentMapping;
+
+                // Navigate into nested properties if available
+                if (currentMap.containsKey("properties")) {
+                    currentProperties = (Map<String, ?>) currentMap.get("properties");
+                }
+                // Handle multifields (e.g., title.raw)
+                else if (currentMap.containsKey("fields") && depth + 1 < fieldParts.length) {
+                    currentProperties = (Map<String, ?>) currentMap.get("fields");
+                }
+                // Return type if found
+                else if (currentMap.containsKey("type")) {
+                    return (String) currentMap.get("type");
+                } else {
+                    return null;
+                }
+            } else {
+                return null;
+            }
+        }
+
+        return null;
+    }
+
+    String getFieldTypeFromCache(String fieldName, Index index) {
+        return fieldTypeMap.getOrDefault(index, new ConcurrentHashMap<>()).get(fieldName);
```

Comment:
Use empty static hashmap here instead of initializing everytime. I will include with my changes